csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedCall.qll

```diff
@@ -0,0 +1,322 @@
+}
+
+/**
+ * IR translation of a direct call to a specific function. Used for both
+ * explicit calls and implicit calls.
+ */
+abstract class TranslatedDirectCall extends TranslatedCall {
+  final override Instruction getFirstCallTargetInstruction() {
+    result = getInstruction(CallTargetTag())
+  }
+
+  final override Instruction getCallTargetResult() { result = getInstruction(CallTargetTag()) }
+
+  override predicate hasInstruction(
+    Opcode opcode, InstructionTag tag, Type resultType, boolean isLValue
+  ) {
+    TranslatedCall.super.hasInstruction(opcode, tag, resultType, isLValue)
+    or
+    tag = CallTargetTag() and
+    opcode instanceof Opcode::FunctionAddress and
+    resultType = expr.getType() and
+    isLValue = true
+  }
+
+  override Instruction getInstructionSuccessor(InstructionTag tag, EdgeKind kind) {
+    result = TranslatedCall.super.getInstructionSuccessor(tag, kind)
+    or
+    tag = CallTargetTag() and
+    kind instanceof GotoEdge and
+    result = getFirstArgumentOrCallInstruction()
+  }
+}
+
+/**
+ * The IR translation of a call to a function.
+ */
+abstract class TranslatedCallExpr extends TranslatedNonConstantExpr, TranslatedCall {
+  override Call expr;
+
+  override Type getCallResultType() { result = getResultType() }
+
+  final override predicate hasArguments() { exists(expr.getArgument(0)) }
+
+  final override TranslatedExpr getQualifier() {
+    expr instanceof QualifiableExpr and
+    result = getTranslatedExpr(expr.(QualifiableExpr).getQualifier()) 
+  }
+
+  final override TranslatedExpr getArgument(int index) {
+    result = getTranslatedExpr(expr.getArgument(index))
+  }
+}
+
+/**
+ * Represents the IR translation of a direct function call.
+ */
+class TranslatedFunctionCall extends TranslatedCallExpr, TranslatedDirectCall {
+  override Call expr;
+
+  override Callable getInstructionFunction(InstructionTag tag) {
+    tag = CallTargetTag() and result = expr.getTarget()
+  }
+
+  override predicate hasReadSideEffect() {
+    not expr.getTarget().(SideEffectFunction).neverReadsMemory()
+  }
+
+  override predicate hasWriteSideEffect() {
+    not expr.getTarget().(SideEffectFunction).neverWritesMemory()
+  }
+}
+
+/**
+ * Represents the IR translation of a call to a constructor.
+ * The target of the call is a newly allocated object whose address, after
+ * the constructor call, address will be passed to a variable declaration.
+ */
+class TranslatedConstructorCall extends TranslatedFunctionCall {
+  TranslatedConstructorCall() { 
+    expr instanceof ObjectCreation or
+    expr instanceof ConstructorInitializer
+  }
+
+  override Instruction getQualifierResult() {
+    // We must retrieve the qualifier from the context the
+    // constructor call happened
+    exists(StructorCallContext context |
+      context = getParent() and
+      result = context.getReceiver()
+    )
+  }
+  
+  override Type getCallResultType() { result instanceof VoidType }
+
+  override predicate hasQualifier() { any() }
```

Comment:
Since they are performed at file level, with no clear entry point, I have skipped them for now.

Reply:
I have also fixed a bug where static callables had the `InitializeThis` instruction.

Reply:
I am not quite sure when those should be called, since they do not have an exact entry point. From what I've read, they are called some time before the first access of the static field (but it is not exactly known when), though if they are not accessed the static constructor is never called.

Reply:
A related question is when or how field initialization should be shown.